pyomo/repn/plugins/nl_writer.py

```diff
@@ -1479,6 +1656,130 @@ def _count_subexpression_occurrences(self):
                 n_subexpressions[0] += 1
         return n_subexpressions
 
+    def _linear_presolve(self, comp_by_linear_var, lcon_by_linear_nnz, var_bounds):
+        eliminated_vars = {}
+        eliminated_cons = set()
+        if not self.config.linear_presolve:
+            return eliminated_cons, eliminated_vars
+
+        fixed_vars = [
+            _id for _id, (lb, ub) in var_bounds.items() if lb == ub and lb is not None
+        ]
+        var_map = self.var_map
+        substitutions_by_linear_var = defaultdict(set)
+        template = self.template
+        one_var = lcon_by_linear_nnz[1]
+        two_var = lcon_by_linear_nnz[2]
+        while 1:
+            if fixed_vars:
+                _id = fixed_vars.pop()
+                a = x = None
+                b, _ = var_bounds[_id]
+                logger.debug("NL presolve: bounds fixed %s := %s", var_map[_id], b)
+                eliminated_vars[_id] = AMPLRepn(b, {}, None)
+            elif one_var:
+                con_id, info = one_var.popitem()
+                expr_info, lb = info
+                _id, coef = expr_info.linear.popitem()
+                # substituting _id with a*x + b
+                a = x = None
+                b = expr_info.const = (lb - expr_info.const) / coef
+                logger.debug("NL presolve: substituting %s := %s", var_map[_id], b)
+                eliminated_vars[_id] = expr_info
+                lb, ub = var_bounds[_id]
+                if (lb is not None and lb - b > TOL) or (
+                    ub is not None and ub - b < -TOL
+                ):
+                    raise InfeasibleConstraintException(
+                        "model contains a trivially infeasible variable "
+                        f"'{var_map[_id].name}' (presolved to a value of "
+                        f"{b} outside bounds [{lb}, {ub}])."
+                    )
+                eliminated_cons.add(con_id)
+            elif two_var:
+                con_id, info = two_var.popitem()
+                expr_info, lb = info
+                _id, coef = expr_info.linear.popitem()
+                id2, coef2 = expr_info.linear.popitem()
+                # For no particularly good reason, we will solve for
+                # (and substitute out) the variable with the smaller
+                # magnitude)
+                if abs(coef2) < abs(coef):
+                    _id, id2 = id2, _id
+                    coef, coef2 = coef2, coef
```

Comment:
I would recommend substituting out the variable with the coefficient closest to 1 for numerical stability (i.e., avoid dividing by large or small numbers).

Reply:
I realize that "closest to 1" is not clear. I think I've measured this using log10 in the past. In other words, we want `log10(coef)` to be close to 0.